framework/validators/Validator.php

```diff
@@ -102,6 +102,10 @@ class Validator extends Component
      */
     public $attributes = [];
     /**
+     * @var array cleaned attribute names with "!" characters in the beginning removed
+     */
+    protected $attributeNames = [];
```

Comment:
if that is ever needed you can add a getter too and use private var in subclass.